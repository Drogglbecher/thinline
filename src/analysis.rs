use error::*;
use glob::glob;
use std::cell::RefMut;
use std::path::PathBuf;

/// A generic trait for source file analysis>
pub trait Anaylsis {
    /// Creates a new Anaylsis instance.
    fn new() -> Self;

    /// Returns the file types which should be analyzed.
    fn file_types(&self) -> &[&str];

    /// Returns the collected project files for anaylsis.
    fn project_files(&self) -> RefMut<Vec<PathBuf>>;

    /// Collect all the sources within the given project dir.
    fn collect_sources(&self, project_dir: &PathBuf, search_dirs: &[&str]) -> Result<()> {
        // Check the given project directory
        if !project_dir.exists() || !project_dir.is_dir() {
            return Err(Error::from(format!(
                "The given project dir '{}' does not exist.",
                project_dir.to_str().ok_or_else(
                    || "Unable to stringify project dir path.",
                )?
            )));
        }

        // Traverse through the files within the specified source directories
        // and store them for analyzing purposes
        for src_dir in search_dirs {
            for ext in self.file_types() {
                for entry in glob(
                    project_dir
                        .join(src_dir)
                        .join("**")
                        .join(ext)
                        .to_str()
                        .unwrap_or("."),
                )?
                {
                    self.project_files().push(entry?);
                }
            }
        }

        Ok(())
    }

    /// Extracts function signatures and comments of thinlines parsed files.
    fn extract_entities(&self) -> Result<()>;
}
