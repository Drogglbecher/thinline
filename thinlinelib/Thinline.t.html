<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=struct.Thinline.html">
</head>
<body>
    <p>Redirecting to <a href="struct.Thinline.html">struct.Thinline.html</a>...</p>
    <script>location.replace("struct.Thinline.html" + location.search + location.hash);</script>
</body>
</html>