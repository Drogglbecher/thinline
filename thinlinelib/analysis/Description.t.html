<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=struct.Description.html">
</head>
<body>
    <p>Redirecting to <a href="struct.Description.html">struct.Description.html</a>...</p>
    <script>location.replace("struct.Description.html" + location.search + location.hash);</script>
</body>
</html>