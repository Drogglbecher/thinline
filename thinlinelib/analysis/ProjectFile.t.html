<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=struct.ProjectFile.html">
</head>
<body>
    <p>Redirecting to <a href="struct.ProjectFile.html">struct.ProjectFile.html</a>...</p>
    <script>location.replace("struct.ProjectFile.html" + location.search + location.hash);</script>
</body>
</html>