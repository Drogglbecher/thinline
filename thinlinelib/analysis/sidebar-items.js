initSidebarItems({"struct":[["Analysis","The analyzer which fulfills parsing and entity extraction tasks."],["Argument","Represents a parsed function argument."],["Description","Represents a parsed entity description."],["Enum","Represents a parsed enum argument."],["Function","Represents a parsed function type."],["ProjectFile","Represents a parsed project file."]]});