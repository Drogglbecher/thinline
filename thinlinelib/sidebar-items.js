initSidebarItems({"mod":[["analysis",""],["entity",""],["language_type",""],["project_parameters",""],["stubs",""],["synthesis",""],["value_parser",""]],"struct":[["Thinline","Global structure representing the `Thinline` lib."]]});