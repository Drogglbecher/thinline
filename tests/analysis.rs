extern crate snapshot;
extern crate thinlinelib;

pub static MULTILINE_COMMENT: &str = "
**this
is
a

multiline
**
//comment";

#[cfg(test)]
mod analysis {

    #[cfg(test)]
    mod argument {
        use thinlinelib::analysis::Argument;

        #[test]
        fn new() {
            let argument = Argument::new("arg", Some("std::string"));

            assert_eq!(argument.name, String::from("arg"));
            assert_eq!(argument.atype, Some(String::from("std::string")));
            assert!(argument.value.is_none());
        }

        #[test]
        fn set_value() {
            let mut argument = Argument::new("arg", Some("std::string"));
            argument.set_value("FirstArg");

            assert_eq!(argument.value, Some(String::from("FirstArg")));
        }
    }

    #[cfg(test)]
    mod function {
        use thinlinelib::analysis::{Argument, Function};
        use MULTILINE_COMMENT;

        #[test]
        fn new() {
            let function = Function::new("fct");

            assert_eq!(function.name, String::from("fct"));
            assert!(function.return_type.is_none());
            assert!(function.arguments.is_empty());
            assert!(function.description.is_none());
        }

        #[test]
        fn set_return_type() {
            let mut fct = Function::new("fct");

            {
                assert!(fct.set_return_type("").is_ok());
                assert_eq!(fct.return_type, None);
            }

            {
                assert!(fct.set_return_type("int").is_ok());
                assert_eq!(fct.return_type, Some(String::from("int")));
            }
        }

        #[test]
        fn set_description() {
            let mut fct = Function::new("fct");

            {
                fct.set_description(MULTILINE_COMMENT);

                assert!(fct.description.is_some());
                let fct_desc = fct.description.unwrap().description;

                assert_eq!(fct_desc.len(), 5);
                assert_eq!(fct_desc[0], "this");
                assert_eq!(fct_desc[1], "is");
                assert_eq!(fct_desc[2], "a");
                assert_eq!(fct_desc[3], "multiline");
                assert_eq!(fct_desc[4], "comment");
            }
        }

        #[test]
        fn set_arguments() {
            let mut fct = Function::new("fct");

            {
                fct.set_arguments(&vec![]);
                assert!(fct.arguments.is_empty());
            }

            {
                let arg1 = Argument::new("arg1", Some("std::string"));
                let arg2 = Argument::new("arg2", Some("std::uint32"));
                fct.set_arguments(&vec![arg1, arg2]);

                assert_eq!(fct.arguments.len(), 2);
                assert_eq!(fct.arguments[0].name, "arg1");
                assert_eq!(fct.arguments[1].name, "arg2");
            }
        }
    }

    #[cfg(test)]
    mod enumeration {
        use thinlinelib::analysis::{Argument, Enum};

        #[test]
        fn new() {
            let enumeration = Enum::new("enum");

            assert_eq!(enumeration.name, String::from("enum"));
            assert!(enumeration.arguments.is_empty());
            assert!(enumeration.etype.is_none());
        }

        #[test]
        fn set_arguments() {
            let mut enumeration = Enum::new("enum");

            {
                enumeration.set_arguments(&vec![]);
                assert!(enumeration.arguments.is_empty());
            }

            {
                let args = vec![
                    Argument::new("Zero", Some("0")),
                    Argument::new("Two", Some("2")),
                ];
                enumeration.set_arguments(&args);

                assert_eq!(enumeration.arguments.len(), 2);
                assert_eq!(enumeration.arguments[0].name, "Zero");
                assert_eq!(enumeration.arguments[1].name, "Two");
            }
        }

        #[test]
        fn push_argument() {
            let mut enumeration = Enum::new("enum");
            assert!(enumeration.arguments.is_empty());

            enumeration.push_argument(Argument::new("arg", Some("uint32")));
            assert_eq!(enumeration.arguments.len(), 1);

            enumeration.push_argument(Argument::new("new_arg", Some("uint64")));
            assert_eq!(enumeration.arguments.len(), 2);
        }
    }

    #[cfg(test)]
    mod c {

        #[cfg(test)]
        mod test_collect_sources {

            #[cfg(test)]
            mod should_succeed {

                use std::path::Path;
                use thinlinelib::analysis::Analysis;
                use thinlinelib::language_type::C;

                #[test]
                fn when_directory_is_valid() {
                    // Given
                    let analysis: Analysis<C> = Analysis::new();

                    // When
                    let c_test_src_path = Path::new("tests").join("testdata").join("c_sources");
                    assert!(
                        analysis
                            .collect_sources(&c_test_src_path, &[String::from(".")])
                            .is_ok()
                    );

                    // Then
                    assert_eq!(analysis.project_files().len(), 5);
                }
            }

            #[cfg(test)]
            mod should_fail {

                use std::path::Path;
                use thinlinelib::analysis::Analysis;
                use thinlinelib::language_type::C;

                #[test]
                fn when_directory_not_existing() {
                    // Given
                    let analysis: Analysis<C> = Analysis::new();

                    // When
                    let c_test_src_path = Path::new("not").join("existing");

                    // Then
                    assert!(
                        analysis
                            .collect_sources(&c_test_src_path, &[String::from(".")])
                            .is_err()
                    );
                }

                #[test]
                fn when_path_is_no_directory() {
                    // Given
                    let analysis: Analysis<C> = Analysis::new();

                    // When
                    let c_test_src_path = Path::new("tests").join("lib.rs");

                    // Then
                    assert!(
                        analysis
                            .collect_sources(&c_test_src_path, &[String::from(".")])
                            .is_err()
                    );
                }
            }

            #[cfg(test)]
            mod extract_entities {
                use snapshot::snapshot;
                use std::path::Path;
                use thinlinelib::analysis::Analysis;
                use thinlinelib::entity::EntityType;
                use thinlinelib::language_type::C;

                fn extract_entities_c() -> Vec<EntityType> {
                    let analysis: Analysis<C> = Analysis::new();
                    let c_test_src_path = Path::new("tests").join("testdata").join("analysis");
                    assert!(
                        analysis
                            .collect_sources(&c_test_src_path, &[String::from(".")])
                            .is_ok()
                    );

                    assert!(analysis.extract_entities().is_ok());

                    let project_files = analysis.project_files();

                    assert_eq!(project_files.len(), 1);

                    let entities = project_files[0].entities();
                    assert_eq!(entities.len(), 1);
                    entities[0].clone().entities
                }

                #[cfg(target_os = "linux")]
                #[snapshot]
                fn extract_entities_linux_c() -> Vec<EntityType> {
                    extract_entities_c()
                }

                #[cfg(target_os = "windows")]
                #[snapshot]
                fn extract_entities_windows_c() -> Vec<EntityType> {
                    extract_entities_c()
                }
            }
        }
    }

    #[cfg(test)]
    mod cpp {

        #[cfg(test)]
        mod test_collect_sources {

            #[cfg(test)]
            mod should_succeed {

                use std::path::Path;
                use thinlinelib::analysis::Analysis;
                use thinlinelib::language_type::Cpp;

                #[test]
                fn when_directory_is_valid() {
                    // Given
                    let analysis: Analysis<Cpp> = Analysis::new();

                    // When
                    let cpp_test_src_path = Path::new("tests").join("testdata").join("cpp_sources");
                    assert!(
                        analysis
                            .collect_sources(&cpp_test_src_path, &[String::from(".")])
                            .is_ok()
                    );

                    // Then
                    assert_eq!(analysis.project_files().len(), 2);
                }
            }

            #[cfg(test)]
            mod should_fail {

                use std::path::Path;
                use thinlinelib::analysis::Analysis;
                use thinlinelib::language_type::Cpp;

                #[test]
                fn when_directory_not_existing() {
                    // Given
                    let analysis: Analysis<Cpp> = Analysis::new();

                    // When
                    let cpp_test_src_path = Path::new("not").join("existing");

                    // Then
                    assert!(
                        analysis
                            .collect_sources(&cpp_test_src_path, &[String::from(".")])
                            .is_err()
                    );
                }

                #[test]
                fn when_path_is_no_directory() {
                    // Given
                    let analysis: Analysis<Cpp> = Analysis::new();

                    // When
                    let cpp_test_src_path = Path::new("tests").join("lib.rs");

                    // Then
                    assert!(
                        analysis
                            .collect_sources(&cpp_test_src_path, &[String::from(".")])
                            .is_err()
                    );
                }
            }

            #[cfg(test)]
            mod extract_entities {
                use snapshot::snapshot;
                use std::path::Path;
                use thinlinelib::analysis::Analysis;
                use thinlinelib::entity::EntityType;
                use thinlinelib::language_type::Cpp;

                fn extract_entities_cpp() -> Vec<EntityType> {
                    let analysis: Analysis<Cpp> = Analysis::new();
                    let cpp_test_src_path = Path::new("tests").join("testdata").join("analysis");
                    assert!(
                        analysis
                            .collect_sources(&cpp_test_src_path, &[String::from(".")])
                            .is_ok()
                    );

                    assert!(analysis.extract_entities().is_ok());

                    let project_files = analysis.project_files();

                    assert_eq!(project_files.len(), 1);

                    let entities = project_files[0].entities();
                    assert_eq!(entities.len(), 1);
                    entities[0].clone().entities
                }

                #[cfg(target_os = "linux")]
                #[snapshot]
                fn extract_entities_linux_cpp() -> Vec<EntityType> {
                    extract_entities_cpp()
                }

                #[cfg(target_os = "windows")]
                #[snapshot]
                fn extract_entities_windows_cpp() -> Vec<EntityType> {
                    extract_entities_cpp()
                }
            }
        }
    }

    #[cfg(test)]
    mod python {

        #[cfg(test)]
        mod collect_sources {

            #[cfg(test)]
            mod should_succeed {

                use std::path::Path;
                use thinlinelib::analysis::Analysis;
                use thinlinelib::language_type::Python;

                #[test]
                fn when_directory_is_valid() {
                    // Given
                    let analysis: Analysis<Python> = Analysis::new();

                    // When
                    let python_test_src_path =
                        Path::new("tests").join("testdata").join("python_sources");
                    assert!(
                        analysis
                            .collect_sources(&python_test_src_path, &[String::from(".")])
                            .is_ok()
                    );

                    // Then
                    assert_eq!(analysis.project_files().len(), 2);
                }
            }

            #[cfg(test)]
            mod should_fail {

                use std::path::Path;
                use thinlinelib::analysis::Analysis;
                use thinlinelib::language_type::Python;

                #[test]
                fn when_directory_not_existing() {
                    // Given
                    let analysis: Analysis<Python> = Analysis::new();

                    // When
                    let python_test_src_path = Path::new("not").join("existing");

                    // Then
                    assert!(
                        analysis
                            .collect_sources(&python_test_src_path, &[String::from(".")])
                            .is_err()
                    );
                }

                #[test]
                fn when_path_is_no_directory() {
                    // Given
                    let analysis: Analysis<Python> = Analysis::new();

                    // When
                    let python_test_src_path = Path::new("tests").join("lib.rs");

                    // Then
                    assert!(
                        analysis
                            .collect_sources(&python_test_src_path, &[String::from(".")])
                            .is_err()
                    );
                }
            }
        }

        #[cfg(test)]
        mod extract_entities {
            use snapshot::snapshot;
            use std::path::Path;
            use thinlinelib::analysis::Analysis;
            use thinlinelib::entity::EntityType;
            use thinlinelib::language_type::Python;

            fn extract_entities_python() -> Vec<EntityType> {
                let analysis: Analysis<Python> = Analysis::new();
                let py_test_src_path = Path::new("tests").join("testdata").join("analysis");
                assert!(
                    analysis
                        .collect_sources(&py_test_src_path, &[String::from(".")])
                        .is_ok()
                );

                assert!(analysis.extract_entities().is_ok());

                let project_files = analysis.project_files();
                assert_eq!(project_files.len(), 1);

                let entities = project_files[0].entities();
                assert_eq!(entities.len(), 1);
                entities[0].clone().entities
            }

            #[cfg(target_os = "linux")]
            #[snapshot]
            fn extract_entities_linux_python() -> Vec<EntityType> {
                extract_entities_python()
            }

            #[cfg(target_os = "windows")]
            #[snapshot]
            fn extract_entities_windows_python() -> Vec<EntityType> {
                extract_entities_python()
            }
        }
    }
}
