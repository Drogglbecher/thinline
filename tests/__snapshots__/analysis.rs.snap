{
  "analysis::analysis::c::test_collect_sources::extract_entities::extract_entities_linux_c": {
    "file": "tests/analysis.rs",
    "module_path": "analysis::analysis::c::test_collect_sources::extract_entities",
    "test_function": "extract_entities_linux_c",
    "recorded_value": [
      {
        "Function": {
          "arguments": [
            {
              "atype": "int",
              "name": "no1",
              "value": null
            },
            {
              "atype": "int",
              "name": "no2",
              "value": null
            }
          ],
          "description": {
            "description": [
              "#TL_TESTCASE(Source1::CheckIfSumWorks)",
              "int test_no = 2;",
              "#TL_EQ[TL_FCT(no1: test_no, no2: 5) => 7]",
              "#TL_EQ[TL_FCT(no1: 5, no2: 2) => 7]",
              "EXPECT_EQ(11, test_int_no1(9, 2));",
              "#!TL_TESTCASE"
            ]
          },
          "name": "test_int_no1",
          "return_type": "int"
        }
      },
      {
        "Function": {
          "arguments": [
            {
              "atype": "const int *const",
              "name": "no1",
              "value": null
            },
            {
              "atype": "const int *const",
              "name": "no2",
              "value": null
            }
          ],
          "description": {
            "description": [
              "#TL_TESTCASE(Source1::TestPtr)",
              "int test_no = 2;",
              "int test_no2 = 5;",
              "#TL_EQ[TL_FCT(no1: &test_no, no2: &test_no2) => 7]",
              "#!TL_TESTCASE"
            ]
          },
          "name": "test_ptr",
          "return_type": "int"
        }
      },
      {
        "Function": {
          "arguments": [],
          "description": {
            "description": [
              "#TL_TESTCASE(Source1::EmptyFct)",
              "#TL_EQ[TL_FCT() => 7]",
              "#TL_NE[TL_FCT() => 4]",
              "#!TL_TESTCASE"
            ]
          },
          "name": "test_empty_fct",
          "return_type": "int"
        }
      },
      {
        "Function": {
          "arguments": [
            {
              "atype": "const int",
              "name": "argc",
              "value": null
            },
            {
              "atype": "char *const []",
              "name": "argv",
              "value": null
            }
          ],
          "description": {
            "description": [
              "This function has parameters, yeah"
            ]
          },
          "name": "main",
          "return_type": "int"
        }
      }
    ]
  },
  "analysis::analysis::c::test_collect_sources::extract_entities::extract_entities_windows_c": {
    "file": "tests\\analysis.rs",
    "module_path": "analysis::analysis::c::test_collect_sources::extract_entities",
    "test_function": "extract_entities_windows_c",
    "recorded_value": [
      {
        "Function": {
          "arguments": [
            {
              "atype": "int",
              "name": "no1",
              "value": null
            },
            {
              "atype": "int",
              "name": "no2",
              "value": null
            }
          ],
          "description": {
            "description": [
              "#TL_TESTCASE(Source1::CheckIfSumWorks)",
              "int test_no = 2;",
              "#TL_EQ[TL_FCT(no1: test_no, no2: 5) => 7]",
              "#TL_EQ[TL_FCT(no1: 5, no2: 2) => 7]",
              "EXPECT_EQ(11, test_int_no1(9, 2));",
              "#!TL_TESTCASE"
            ]
          },
          "name": "test_int_no1",
          "return_type": "int"
        }
      },
      {
        "Function": {
          "arguments": [
            {
              "atype": "const int *const",
              "name": "no1",
              "value": null
            },
            {
              "atype": "const int *const",
              "name": "no2",
              "value": null
            }
          ],
          "description": {
            "description": [
              "#TL_TESTCASE(Source1::TestPtr)",
              "int test_no = 2;",
              "int test_no2 = 5;",
              "#TL_EQ[TL_FCT(no1: &test_no, no2: &test_no2) => 7]",
              "#!TL_TESTCASE"
            ]
          },
          "name": "test_ptr",
          "return_type": "int"
        }
      },
      {
        "Function": {
          "arguments": [],
          "description": {
            "description": [
              "#TL_TESTCASE(Source1::EmptyFct)",
              "#TL_EQ[TL_FCT() => 7]",
              "#TL_NE[TL_FCT() => 4]",
              "#!TL_TESTCASE"
            ]
          },
          "name": "test_empty_fct",
          "return_type": "int"
        }
      },
      {
        "Function": {
          "arguments": [
            {
              "atype": "const int",
              "name": "argc",
              "value": null
            },
            {
              "atype": "char *const []",
              "name": "argv",
              "value": null
            }
          ],
          "description": {
            "description": [
              "This function has parameters, yeah"
            ]
          },
          "name": "main",
          "return_type": "int"
        }
      }
    ]
  },
  "analysis::analysis::cpp::test_collect_sources::extract_entities::extract_entities_linux_cpp": {
    "file": "tests/analysis.rs",
    "module_path": "analysis::analysis::cpp::test_collect_sources::extract_entities",
    "test_function": "extract_entities_linux_cpp",
    "recorded_value": [
      {
        "Entity": {
          "description": null,
          "entities": [
            {
              "Entity": {
                "description": null,
                "entities": [
                  {
                    "Function": {
                      "arguments": [
                        {
                          "atype": "unsigned int",
                          "name": "no1",
                          "value": null
                        },
                        {
                          "atype": "unsigned int",
                          "name": "no2",
                          "value": null
                        }
                      ],
                      "description": {
                        "description": [
                          "#TL_TESTCASE(c1::AddTwoNumbers)",
                          "unsigned int no1 = 5;",
                          "#TL_EQ[this->class_inst->TL_FCT(no1: no1, no2: 10) => 15]",
                          "#TL_LT[this->class_inst->TL_FCT(no1: no1, no2: 10) => 30]",
                          "#!TL_TESTCASE"
                        ]
                      },
                      "name": "add_two_numbers",
                      "return_type": "unsigned int"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "c1",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "~c1",
                      "return_type": "void"
                    }
                  }
                ],
                "name": "c1"
              }
            },
            {
              "Entity": {
                "description": null,
                "entities": [
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "c2",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "~c2",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [
                        {
                          "atype": "unsigned int",
                          "name": "no1",
                          "value": null
                        },
                        {
                          "atype": "unsigned int",
                          "name": "no2",
                          "value": null
                        },
                        {
                          "atype": "unsigned int",
                          "name": "no3",
                          "value": null
                        }
                      ],
                      "description": {
                        "description": [
                          "#TL_TESTCASE(c2::AddThreeNumbers)",
                          "unsigned int no1 = 5;",
                          "unsigned int no2 = 10;",
                          "#TL_EQ[this->class_inst->TL_FCT(no1: no1, no2: no2, no3: 5) => 20]",
                          "#!TL_TESTCASE"
                        ]
                      },
                      "name": "add_three_numbers",
                      "return_type": "unsigned int"
                    }
                  }
                ],
                "name": "c2"
              }
            }
          ],
          "name": "ns1"
        }
      },
      {
        "Entity": {
          "description": null,
          "entities": [
            {
              "Entity": {
                "description": null,
                "entities": [
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "c3",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "~c3",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "return5",
                      "return_type": "unsigned int"
                    }
                  }
                ],
                "name": "c3"
              }
            }
          ],
          "name": "ns2"
        }
      }
    ]
  },
  "analysis::analysis::cpp::test_collect_sources::extract_entities::extract_entities_windows_cpp": {
    "file": "tests\\analysis.rs",
    "module_path": "analysis::analysis::cpp::test_collect_sources::extract_entities",
    "test_function": "extract_entities_windows_cpp",
    "recorded_value": [
      {
        "Entity": {
          "description": null,
          "entities": [
            {
              "Entity": {
                "description": null,
                "entities": [
                  {
                    "Function": {
                      "arguments": [
                        {
                          "atype": "unsigned int",
                          "name": "no1",
                          "value": null
                        },
                        {
                          "atype": "unsigned int",
                          "name": "no2",
                          "value": null
                        }
                      ],
                      "description": {
                        "description": [
                          "#TL_TESTCASE(c1::AddTwoNumbers)",
                          "unsigned int no1 = 5;",
                          "#TL_EQ[this->class_inst->TL_FCT(no1: no1, no2: 10) => 15]",
                          "#TL_LT[this->class_inst->TL_FCT(no1: no1, no2: 10) => 30]",
                          "#!TL_TESTCASE"
                        ]
                      },
                      "name": "add_two_numbers",
                      "return_type": "unsigned int"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "c1",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "~c1",
                      "return_type": "void"
                    }
                  }
                ],
                "name": "c1"
              }
            },
            {
              "Entity": {
                "description": null,
                "entities": [
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "c2",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "~c2",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [
                        {
                          "atype": "unsigned int",
                          "name": "no1",
                          "value": null
                        },
                        {
                          "atype": "unsigned int",
                          "name": "no2",
                          "value": null
                        },
                        {
                          "atype": "unsigned int",
                          "name": "no3",
                          "value": null
                        }
                      ],
                      "description": {
                        "description": [
                          "#TL_TESTCASE(c2::AddThreeNumbers)",
                          "unsigned int no1 = 5;",
                          "unsigned int no2 = 10;",
                          "#TL_EQ[this->class_inst->TL_FCT(no1: no1, no2: no2, no3: 5) => 20]",
                          "#!TL_TESTCASE"
                        ]
                      },
                      "name": "add_three_numbers",
                      "return_type": "unsigned int"
                    }
                  }
                ],
                "name": "c2"
              }
            }
          ],
          "name": "ns1"
        }
      },
      {
        "Entity": {
          "description": null,
          "entities": [
            {
              "Entity": {
                "description": null,
                "entities": [
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "c3",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "~c3",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "return5",
                      "return_type": "unsigned int"
                    }
                  }
                ],
                "name": "c3"
              }
            }
          ],
          "name": "ns2"
        }
      }
    ]
  },
  "analysis::analysis::python::extract_entities::extract_entities_linux_python": {
    "file": "tests/analysis.rs",
    "module_path": "analysis::analysis::python::extract_entities",
    "test_function": "extract_entities_linux_python",
    "recorded_value": [
      {
        "Function": {
          "arguments": [
            {
              "atype": null,
              "name": "no1",
              "value": null
            },
            {
              "atype": null,
              "name": "no2",
              "value": null
            }
          ],
          "description": {
            "description": [
              "#TL_TESTCASE(check_if_sum_works)",
              "int test_no = 2;",
              "#TL_EQ[TL_FCT(no1: test_no, no2: 5) => 7]",
              "EXPECT_EQ(11, test_int_no1(9, 2));",
              "#!TL_TESTCASE"
            ]
          },
          "name": "test_int_no1",
          "return_type": null
        }
      },
      {
        "Entity": {
          "description": null,
          "entities": [
            {
              "Function": {
                "arguments": [
                  {
                    "atype": null,
                    "name": "float1",
                    "value": null
                  },
                  {
                    "atype": null,
                    "name": "float2",
                    "value": null
                  }
                ],
                "description": {
                  "description": [
                    "#TL_TESTCASE(check_if_sum_works)",
                    "#TL_EQ[TL_FCT(float1: 4.2, float2: 3.2) => 7.4]",
                    "#!TL_TESTCASE"
                  ]
                },
                "name": "test_float",
                "return_type": null
              }
            },
            {
              "Function": {
                "arguments": [],
                "description": null,
                "name": "test_nodoc",
                "return_type": null
              }
            },
            {
              "Function": {
                "arguments": [
                  {
                    "atype": null,
                    "name": "str1",
                    "value": null
                  },
                  {
                    "atype": null,
                    "name": "str2",
                    "value": null
                  }
                ],
                "description": {
                  "description": [
                    "#TL_TESTCASE(check_if_str_concat_works)",
                    "#TL_EQ[TL_FCT(str1: 'bla', str2: 'blub') => 'blablub']",
                    "#!TL_TESTCASE"
                  ]
                },
                "name": "test_str",
                "return_type": null
              }
            }
          ],
          "name": "class1"
        }
      }
    ]
  },
  "analysis::analysis::python::extract_entities::extract_entities_windows_python": {
    "file": "tests\\analysis.rs",
    "module_path": "analysis::analysis::python::extract_entities",
    "test_function": "extract_entities_windows_python",
    "recorded_value": [
      {
        "Function": {
          "arguments": [
            {
              "atype": null,
              "name": "no1",
              "value": null
            },
            {
              "atype": null,
              "name": "no2",
              "value": null
            }
          ],
          "description": {
            "description": [
              "#TL_TESTCASE(check_if_sum_works)",
              "int test_no = 2;",
              "#TL_EQ[TL_FCT(no1: test_no, no2: 5) => 7]",
              "EXPECT_EQ(11, test_int_no1(9, 2));",
              "#!TL_TESTCASE"
            ]
          },
          "name": "test_int_no1",
          "return_type": null
        }
      },
      {
        "Entity": {
          "description": null,
          "entities": [
            {
              "Function": {
                "arguments": [
                  {
                    "atype": null,
                    "name": "float1",
                    "value": null
                  },
                  {
                    "atype": null,
                    "name": "float2",
                    "value": null
                  }
                ],
                "description": {
                  "description": [
                    "#TL_TESTCASE(check_if_sum_works)",
                    "#TL_EQ[TL_FCT(float1: 4.2, float2: 3.2) => 7.4]",
                    "#!TL_TESTCASE"
                  ]
                },
                "name": "test_float",
                "return_type": null
              }
            },
            {
              "Function": {
                "arguments": [],
                "description": null,
                "name": "test_nodoc",
                "return_type": null
              }
            },
            {
              "Function": {
                "arguments": [
                  {
                    "atype": null,
                    "name": "str1",
                    "value": null
                  },
                  {
                    "atype": null,
                    "name": "str2",
                    "value": null
                  }
                ],
                "description": {
                  "description": [
                    "#TL_TESTCASE(check_if_str_concat_works)",
                    "#TL_EQ[TL_FCT(str1: 'bla', str2: 'blub') => 'blablub']",
                    "#!TL_TESTCASE"
                  ]
                },
                "name": "test_str",
                "return_type": null
              }
            }
          ],
          "name": "class1"
        }
      }
    ]
  }
}
